src/main/java/redis/clients/jedis/JedisSlotBasedConnectionHandler.java

```diff
@@ -44,9 +43,12 @@ public Jedis getConnection() {
         if (result.equalsIgnoreCase("pong")) return jedis;
 
         pool.returnBrokenResource(jedis);
-      } catch (JedisConnectionException ex) {
+      } catch (JedisException ex) {
         if (jedis != null) {
-          pool.returnBrokenResource(jedis);
+          try {
```

Comment:
@marcosnils 
When JedisDataException is occurred, the connection is not marked to be broken. So jedis.close() doesn't work on this situation.
And if we want to consume any Exceptions during returning, we can even define it to consume Throwable since it is more safer.